### PR TITLE
KEYCLOAK-2045 Add support for flexible Validation

### PR DESCRIPTION
**Pull request by thomasdarimont**
This proposal shall serve as a more flexible foundation for user profile validations and much more.

This is a squashed and cleaned-up version of my experimental branch [poc/validation-spi](https://github.com/keycloak/keycloak/compare/master...thomasdarimont:poc/validation-spi?expand=1).

See discussion on mailing list: https://groups.google.com/g/keycloak-dev/c/-XjQu7rn56s

Some API examples can be found in the [DefaultValidatorProviderTest](https://github.com/keycloak/keycloak/blob/7c9ca79ef52072ea12602705d53b346addd3535d/services/src/test/java/org/keycloak/services/validation/DefaultValidatorProviderTest.java)

Additionally some integration examples can be seen here: 
- [UpdateProfile.java](https://github.com/thomasdarimont/keycloak/blob/863f0a89a88e5ac0ee179e786922deebc31fbfc9/services/src/main/java/org/keycloak/authentication/requiredactions/UpdateProfile.java#L71)
- [Validation.java](https://github.com/thomasdarimont/keycloak/blob/863f0a89a88e5ac0ee179e786922deebc31fbfc9/services/src/main/java/org/keycloak/services/validation/Validation.java#L99)

### CODE REVIEW

**Review by pedroigor — COMMENTED** [review]
@thomasdarimont Nice stuff! 

I don't have much to add but a few implementation points like those for loops to lookup keys that could maybe be improved with a map.

The other thing would be the correct place for this new SPI, if not in `server-spi-private`. At least for now and while we get more feedback from people using it.

**Comment by thomasdarimont** [comment]
@pedroigor thanks for the quick review. Okay, so I'm going to move everything under server-spi-private for now and address the loop idea.

**Comment by thomasdarimont** [comment]
@thomasdarimont I adapted the PR according to your changes.

I moved the interfaces from server-spi to server-spi-private and refactored the key lookups for ValidationKey and ValidationContextKey. I just noticed that the `org.keycloak.validation.ClientValidationSPI` could get replaced with the generic validation SPI.

**Comment by thomasdarimont** [comment]
To ease access to the validation API, we could also add a method to `KeycloakSession` like:

```
/**
 * Provides access to a built-in ValidatorProvider which can validate values according to various validation rules.
 */
ValidatorProvider validation();
```

File: server-spi-private/src/main/java/org/keycloak/validation/ValidatorProvider.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    default ValidationResult validate(ValidationContext context, Object value, ValidationKey... keys) {
+        return validate(context, value, new LinkedHashSet<>(Arrays.asList(keys)));
+    }
+
+    default ValidationResult validate(ValidationContext context, Object value, ValidationKey key) {
+        return validate(context, value, Collections.singleton(key));
+    }
+
+    default ValidationResult validate(ValidationContext context, Object value) {
```
Comment (thomasdarimont) [inline server-spi-private/src/main/java/org/keycloak/validation/ValidatorProvider.java:50]: We could also offer a variant like:
```
 ValidationResult validate(Object value)
```
which could create a ValidationContext internally which extracts information like the current, realm, client, etc. from the `KeycloakSession.getContext()`

File: server-spi-private/src/main/java/org/keycloak/validation/ConditionalValidation.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ */
+package org.keycloak.validation;
+
+/**
+ * {@link ConditionalValidation} allows to delegate to a given validation while customizing the checks for
+ * enablement and context-specific support.
+ */
+public class ConditionalValidation implements Validation {
+
+    private final Validation delegate;
```
Comment (thomasdarimont) [inline server-spi-private/src/main/java/org/keycloak/validation/ConditionalValidation.java:25]: delegate could be renamed to validation

File: server-spi-private/src/main/java/org/keycloak/validation/ValidationContext.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    // user registration, user profile update, client registration, realm creation
+    private final ValidationContextKey contextKey;
+
+    // additional context specific attributes
+    private final Map<String, Object> attributes;
+
+    /**
+     * Holds the current {@link RealmModel}
+     */
+    private final RealmModel realm;
```
Comment (thomasdarimont) [inline server-spi-private/src/main/java/org/keycloak/validation/ValidationContext.java:41]: could be extracted from the `KeycloakSession`

File: server-spi-private/src/main/java/org/keycloak/validation/ValidationContext.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+    /**
+     * Holds the current {@link UserModel}
+     */
+    private final UserModel user;
+
+    /**
+     * Holds the current {@link ClientModel}
+     */
+    private final ClientModel client;
```
Comment (thomasdarimont) [inline server-spi-private/src/main/java/org/keycloak/validation/ValidationContext.java:51]: could be extracted from the `KeycloakSession`

File: server-spi-private/src/main/java/org/keycloak/validation/ValidationContext.java [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+    /**
+     * @param realm
+     * @param contextKey
+     * @param attributes
+     * @param user
+     * @param client
+     * @param failFast
+     */
+    public ValidationContext(ValidationContextKey contextKey, Map<String, Object> attributes, RealmModel realm, UserModel user, ClientModel client, boolean failFast, ValidationActionKey actionKey) {
```
Comment (thomasdarimont) [inline server-spi-private/src/main/java/org/keycloak/validation/ValidationContext.java:89]: We could add a ctor variant that accepts a KeycloakSession, with this we could access Realm and Client information via `session.getContext()`

**Comment by thomasdarimont** [comment]
Just rebased this on current master

**Comment by pedroigor** [comment]
@thomasdarimont I liked it very much. Great work. It LGTM and I'm OK to merge it.

One thing though, what do you think about this construct?

```
ValidationContext context = validator.forUserProfile();

context.withFailFast(true);

ValidationResult result = context.validate("user@kc.org", ValidationKey.USER_EMAIL);
```

Which could possibly be chained as:

```
validator.forUserProfile().withFailFast().validate("user@kc.org", ValidationKey.USER_EMAIL);
```

Or this one to validate everything on a user:

```
validator.forUserProfile().withFailFast().validate(user);
```

Considering that the built-in context keys are usually what you'll be validating maybe we could make the usage simpler by not forcing client code to explicitly use constants, what I think makes code more readable.

**Comment by thomasdarimont** [comment]
Hi @pedroigor,

Thank you for your feedback! Of course we could add some convenience methods but I'd add those to a dedicated class instead of the default Validator interface.

Btw. I just noticed that a stripped down version of a validation framework was just added to keycloak in 11.0.3
https://github.com/keycloak/keycloak/commit/f660adc9f820c28de7dce42b6fff2c3bf39684fb

With this PR I think there are now at least 4 different ways to do validation in Keycloak.
- Java EE Bean Validation
- The old ValidatorClass
- the new Validator provider added with 11.0.3
- this PR.

I need to check again whether those can be combined, or whether the new functionality partially supersedes the Validation implementation of this PR.

Cheers,
Thomas

**Comment by pedroigor** [comment]
As per the different validation approaches, that is one of my main concerns right now :)

So far the validation we had was specific for clients but now it is changing the scope and becoming a really overlap with your changes and make things even more interesting (or not). See https://github.com/keycloak/keycloak/pull/7586.

Regarding the convenience methods, my main point is related to not making this API so generic but tied with our requirements and use cases. Thus the proposed convenience methods from a single place instead of spreading them across different classes.

**Comment by pedroigor** [comment]
@thomasdarimont How we are going to proceed with this one?

**Comment by thomasdarimont** [comment]
@pedroigor sorry for the late reply I was busy quite this week... as discussed. I'll catch up with the latest design document and create a smaller (but extensible) version of this PR without all the bells and whistles.

We can then close this PR.

**Comment by pedroigor** [comment]
Thanks.

**Pull request closed without merging**